### PR TITLE
Intel MDS-II: correct dump for PIO & regression fixes

### PR DESCRIPTION
**Pull request by fulivi**
Hi,
this PR is for a couple of changes to imds2 driver.

- The correct dump for PIO controller has been added
- A regression introduced by 4287970 has been fixed. The iSBC202 floppy controller broke because of this regression.

Thanks.
--F.Ulivi

### CODE REVIEW

**Comment by fulivi** [comment]
While I was at it I optimized the performance of isbc202 by changing devcb's to devdelegates..

File: src/devices/bus/multibus/multibus.h [hunk 1]
```diff
@@ -87,6 +87,12 @@ class multibus_device
 	template <unsigned I> auto int_callback() { return m_int_cb[I].bind(); }
 	template <unsigned I> void int_w(int state) { m_int_cb[I](state); }
 
+	// wait signal (split for read/write operations)
+	auto wait_io_rd_cb() { return m_wait_rd_cb.bind(); }
+	auto wait_io_wr_cb() { return m_wait_wr_cb.bind(); }
+	void set_wait_io_rd(int state) { m_wait_rd_cb(state); }
+	void set_wait_io_wr(int state) { m_wait_wr_cb(state); }
```
Comment (cuavas) [inline src/devices/bus/multibus/multibus.h:94]: You’re mixing live signal methods with configuration methods.  This leads to confusion – they should be clearly separated.

File: src/devices/bus/multibus/isbc202.h [hunk 2]
```diff
@@ -31,7 +31,7 @@ class isbc202_device : public cpu_device,
 	uint8_t io_r(address_space &space, offs_t offset);
 	void io_w(address_space &space, offs_t offset, uint8_t data);
 
-	DECLARE_WRITE_LINE_MEMBER(co_w);
+	void co_w(bool state);
```
Comment (cuavas) [inline src/devices/bus/multibus/isbc202.h:34]: Why the change?  We still conventionally use `DECLARE_WRITE_LINE_MEMBER` for lines like this.

File: src/devices/bus/multibus/isbc202.cpp [hunk 3]
```diff
@@ -497,34 +497,34 @@ void isbc202_device::device_add_mconfig(machine_config &config)
...
-	m_cpes[ 2 ]->ibus_r().set([this]() { return ibus_r() >> 4; });
-	m_cpes[ 3 ]->ibus_r().set([this]() { return ibus_r() >> 6; });
+	m_cpes[ 0 ]->set_ibus_r_cb([this]() { return ibus_r(); } , "");
+	m_cpes[ 1 ]->set_ibus_r_cb([this]() { return ibus_r() >> 2; } , "");
+	m_cpes[ 2 ]->set_ibus_r_cb([this]() { return ibus_r() >> 4; } , "");
+	m_cpes[ 3 ]->set_ibus_r_cb([this]() { return ibus_r() >> 6; } , "");
 
 	// Connect SX input
-	m_mcu->sx_r().set([this]() { return m_microcode_addr & 0xf; });
+	m_mcu->set_sx_r_cb([this]() { return m_microcode_addr & 0xf; } , "");
```
Comment (cuavas) [inline src/devices/bus/multibus/isbc202.cpp:525]: Isn’t this a place you should be using the `NAME` macro on the lambdas so you don’t get an empty string in diagnostic messages if something goes wrong in the callback?

File: src/devices/machine/i3001.h [hunk 4]
```diff
@@ -54,7 +54,7 @@ class i3001_device : public device_t
 	void fc_w(uint8_t fc);
 
 	// Write Flag Input
-	DECLARE_WRITE_LINE_MEMBER(fi_w) { m_fi = state; }
+	void fi_w(bool state) { m_fi = state; }
```
Comment (cuavas) [inline src/devices/machine/i3001.h:57]: Once again, this goes against conventions.

File: src/devices/machine/i3001.h [hunk 5]
```diff
@@ -64,11 +64,11 @@ class i3001_device : public device_t
...
 
 	// Output callbacks
-	auto fo_w() { return m_fo_handler.bind(); }
+	template <typename... T> void set_fo_w_cb(T &&... args) { m_fo_handler.set(std::forward<T>(args)...); }
 
 	// Input callbacks
-	auto px_r() { return m_px_handler.bind(); }
-	auto sx_r() { return m_sx_handler.bind(); }
+	template <typename... T> void set_px_r_cb(T &&... args) { m_px_handler.set(std::forward<T>(args)...); }
+	template <typename... T> void set_sx_r_cb(T &&... args) { m_sx_handler.set(std::forward<T>(args)...); }
```
Comment (cuavas) [inline src/devices/machine/i3001.h:71]: These are configuration methods between live methods.  Please put all the configuration methods at the top of the class declaration.

File: src/devices/machine/i3001.h [hunk 6]
```diff
@@ -80,9 +80,9 @@ class i3001_device : public device_t
 	virtual void device_start() override;
 
 private:
-	devcb_write_line m_fo_handler;
-	devcb_read8 m_px_handler;
-	devcb_read8 m_sx_handler;
+	device_delegate<void (bool)> m_fo_handler;
+	device_delegate<uint8_t (void)> m_px_handler;
+	device_delegate<uint8_t (void)> m_sx_handler;
```
Comment (cuavas) [inline src/devices/machine/i3001.h:85]: Please use empty parentheses for function argument lists with no arguments, not C-style `(void)`.

File: src/devices/machine/i3002.h [hunk 7]
```diff
@@ -68,10 +68,10 @@ class i3002_device : public device_t
 	void fc_kbus_w(uint8_t fc , uint8_t k);
 
 	// Write Carry Input
-	DECLARE_WRITE_LINE_MEMBER(ci_w) { m_ci = state != 0; }
+	void ci_w(bool state) { m_ci = state; }
 
 	// Write Left Input
-	DECLARE_WRITE_LINE_MEMBER(li_w) { m_li = state != 0; }
+	void li_w(bool state) { m_li = state; }
```
Comment (cuavas) [inline src/devices/machine/i3002.h:74]: Please don’t go changing things like this to go against conventions.

File: src/devices/machine/i3002.h [hunk 8]
```diff
@@ -80,12 +80,12 @@ class i3002_device : public device_t
...
-	auto co_w() { return m_co_handler.bind(); }
-	auto ro_w() { return m_ro_handler.bind(); }
+	template <typename... T> void set_co_w_cb(T &&... args) { m_co_handler.set(std::forward<T>(args)...); }
+	template <typename... T> void set_ro_w_cb(T &&... args) { m_ro_handler.set(std::forward<T>(args)...); }
 
 	// Input bus callbacks
-	auto ibus_r() { return m_ibus_handler.bind(); }
-	auto mbus_r() { return m_mbus_handler.bind(); }
+	template <typename... T> void set_ibus_r_cb(T &&... args) { m_ibus_handler.set(std::forward<T>(args)...); }
+	template <typename... T> void set_mbus_r_cb(T &&... args) { m_mbus_handler.set(std::forward<T>(args)...); }
```
Comment (cuavas) [inline src/devices/machine/i3002.h:88]: Same comment about mixing live and configuration methods.

File: src/devices/machine/i3002.h [hunk 9]
```diff
@@ -106,10 +106,10 @@ class i3002_device : public device_t
...
 
 private:
-	devcb_write_line m_co_handler;
-	devcb_write_line m_ro_handler;
-	devcb_read8 m_ibus_handler;
-	devcb_read8 m_mbus_handler;
+	device_delegate<void (bool)> m_co_handler;
+	device_delegate<void (bool)> m_ro_handler;
+	device_delegate<uint8_t (void)> m_ibus_handler;
+	device_delegate<uint8_t (void)> m_mbus_handler;
```
Comment (cuavas) [inline src/devices/machine/i3002.h:112]: Same comment about C++ conventions.

**Review by cuavas — CHANGES_REQUESTED** [review]
Do all the things you’ve changed from device callbacks to delegates get called frequently enough to justify the reduced flexibility?

File: src/devices/bus/multibus/isbc202.cpp (hunk 3 above)
Comment (fulivi) [reply src/devices/bus/multibus/isbc202.cpp]: Ah, didn't know that NAME could (should) be applied to lambdas too..

**Comment by fulivi** [comment]
To answer your question about whether it was worth to change from devcb to delegates or not: yes, I think it was definitely worth it. On my ancient i5 machine the imds2+isbc202 combo jumped from 80% to 200% speed. After all those callbacks are called multiple times during each cycle of the bit-slice processor (it runs at about 2.5 MHz).
As a last note: I took inspiration from your cc401fc7742 change on the HP Hybrid CPU.

**Comment by fulivi** [comment]
Here it is... rebased after source-reorg-aggedon

**Comment by fulivi** [comment]
Hi, do I still have any hope to see this PR merged (maybe with some more changes) or should I close it and keep the changes off-tree, please?

File: src/devices/bus/multibus/multibus.h [hunk 10]
```diff
@@ -87,6 +87,14 @@ class multibus_device
 	template <unsigned I> auto int_callback() { return m_int_cb[I].bind(); }
 	template <unsigned I> void int_w(int state) { m_int_cb[I](state); }
 
+	// wait signal callbacks (split for read/write operations)
+	auto wait_io_rd_cb() { return m_wait_rd_cb.bind(); }
+	auto wait_io_wr_cb() { return m_wait_wr_cb.bind(); }
+
```
Comment (pmackinlay) [inline src/devices/bus/multibus/multibus.h:93]: I object to these callbacks as they don't correspond to the actual Multibus specification. Multibus uses a transfer acknowledge (/XACK) signal to indicate when a read or write transfer is complete (or that a transfer is being timed out). If the iSBC 202 hardware absolutely requires the bus master to be stalled until the slave responds, a more accurate approach to emulating the bus is necessary.

One suggestion is for two-phase bus read/write logic to be implemented in the Multibus bus emulation, preferably also readily supporting hardware that doesn't depend on precise /XACK implementation.

**Comment by fulivi** [comment]
Hi, I know that multibus uses the XACK/ signal to close bus transactions.
It's exactly through this mechanism that iSBC202 manages to stall the CPU
when it's doing I/O to some ports (status port in particular). In case of
status port, for example, bus is released and "IN" operation allowed to
complete when bit-slice processor on 202 has prepared the status byte.
I can understand why you think my implementation is not very elegant though
it works with 0 cost when the stalling action is not needed (I.e. when the
card issues normal XACKs).
I don't know what you mean by "two-phase logic ", in order to rewrite this
mechanism in an alternative, more elegant and accurate way I definitely
need your help, please.
Thanks.

Il Lun 18 Lug 2022, 05:06 Patrick Mackinlay ***@***.***> ha
scritto:

**Comment by fulivi** [comment]
Any suggestion, please? How would you implement this CPU-stalling function if you were to do it yourself? Just the general idea, I think I can fill the details.

**Comment by pmackinlay** [comment]
While it's not ideal, I will withdraw my objection if you change the `wait_io_rd` and `wait_io_wr` lines to a single line named `xack` (and implement with proper polarity such that xack_w(0) means it's asserted), then use it within your driver and devices the same as you do now. This at least will match the bus specification slightly more faithfully than inventing two completely fictitious signals.

Although I haven't looked at the code the CPU is executing, hopefully rewinding the PC by 2 bytes will work for both read and write operations without any other side effects, eliminating the need for two separate lines?

I would also prefer to see the processor card itself refactored as a Multibus card instead of built into the driver as you have it now, however I won't hold this PR up for that reason.

**Comment by fulivi** [comment]
Thanks. Are these changes in line with your suggestion, please?
